File: ts/tests/graph.test.ts
import { describe, it, expect } from "vitest";
import { Graph, Query, GraphStateError } from "../src/graph.js";
import { InMemoryStore } from "../src/store.js";
import { InMemoryActorStore, Actor, ActorType, ActorStatus } from "../src/actor.js";
import { NoopSigner } from "../src/event.js";
import {
  ActorId, ConversationId, EventType, PublicKey, SubscriptionPattern,
} from "../src/types.js";
import { DefaultTrustModel } from "../src/trust.js";
import { DefaultAuthorityChain, AuthorityLevel } from "../src/authority.js";

function makeGraph(options?: Parameters<typeof Graph.prototype.constructor>[2]) {
  return new Graph(new InMemoryStore(), new InMemoryActorStore(), options);
}

function makeActor(name: string): { actorId: ActorId; publicKey: PublicKey } {
  const bytes = new Uint8Array(32);
  for (let i = 0; i < name.length && i < 32; i++) {
    bytes[i] = name.charCodeAt(i);
  }
  return { actorId: new ActorId(name), publicKey: new PublicKey(bytes) };
}

describe("Graph", () => {
  it("bootstrap creates genesis event", () => {
    const g = makeGraph();
    const actorId = new ActorId("system");
    const ev = g.bootstrap(actorId);

    expect(ev.type.value).toBe("system.bootstrapped");
    expect(ev.source.value).toBe("system");
    expect(g.store.count()).toBe(1);
  });

  it("start and close lifecycle", () => {
    const g = makeGraph();
    g.bootstrap(new ActorId("system"));
    g.start();
    // Should be able to query after start
    const q = g.query();
    expect(q.eventCount()).toBe(1);

    g.close();
    // After close, operations throw
    expect(() => g.query()).toThrow(GraphStateError);
  });

  it("record requires started graph", () => {
    const g = makeGraph();
    g.bootstrap(new ActorId("system"));
    // Not started yet
    expect(() =>
      g.record(
        new EventType("test.event"), new ActorId("alice"), {},
        [], new ConversationId("conv_1"),
      ),
    ).toThrow(GraphStateError);
  });

  it("record requires non-empty store", () => {
    const g = makeGraph();
    g.start();
    expect(() =>
      g.record(
        new EventType("test.event"), new ActorId("alice"), {},
        [], new ConversationId("conv_1"),
      ),
    ).toThrow(GraphStateError);
  });

  it("record appends and publishes event", () => {
    const g = makeGraph();
    const boot = g.bootstrap(new ActorId("system"));
    g.start();

    const published: string[] = [];
    g.bus.subscribe(new SubscriptionPattern("*"), (ev) => {
      published.push(ev.type.value);
    });

    const ev = g.record(
      new EventType("trust.updated"), new ActorId("alice"),
      { current: 0.5 }, [boot.id], new ConversationId("conv_1"),
    );

    expect(ev.type.value).toBe("trust.updated");
    expect(g.store.count()).toBe(2);
    expect(published).toContain("trust.updated");
  });

  it("bootstrap publishes on bus", () => {
    const g = makeGraph();
    const published: string[] = [];
    g.bus.subscribe(new SubscriptionPattern("system.*"), (ev) => {
      published.push(ev.type.value);
    });

    g.bootstrap(new ActorId("system"));
    expect(published).toEqual(["system.bootstrapped"]);
  });

  it("evaluate delegates to authority chain", () => {
    const trustModel = new DefaultTrustModel();
    const chain = new DefaultAuthorityChain(trustModel);
    const g = makeGraph({ trustModel, authorityChain: chain });
    g.bootstrap(new ActorId("system"));
    g.start();

    const { publicKey } = makeActor("bob");
    const actor = g.actorStore.register(publicKey, "Bob", ActorType.Human);

    const result = g.evaluate(actor, "some.action");
    expect(result.level).toBe(AuthorityLevel.Notification);
    expect(result.chain.length).toBeGreaterThan(0);
  });

  it("query.recent returns events newest first", () => {
    const g = makeGraph();
    const boot = g.bootstrap(new ActorId("system"));
    g.start();

    const e1 = g.record(
      new EventType("test.one"), new ActorId("alice"), {},
      [boot.id], new ConversationId("conv_1"),
    );
    const e2 = g.record(
      new EventType("test.two"), new ActorId("alice"), {},
      [e1.id], new ConversationId("conv_1"),
    );

    const recent = g.query().recent(10);
    expect(recent.length).toBe(3);
    expect(recent[0].id.value).toBe(e2.id.value);
    expect(recent[2].id.value).toBe(boot.id.value);
  });

  it("query.byType filters by event type", () => {
    const g = makeGraph();
    const boot = g.bootstrap(new ActorId("system"));
    g.start();

    g.record(
      new EventType("test.alpha"), new ActorId("alice"), {},
      [boot.id], new ConversationId("conv_1"),
    );
    g.record(
      new EventType("test.beta"), new ActorId("alice"), {},
      [boot.id], new ConversationId("conv_1"),
    );

    const results = g.query().byType(new EventType("test.alpha"), 10);
    expect(results.length).toBe(1);
    expect(results[0].type.value).toBe("test.alpha");
  });

  it("query.bySource filters by actor", () => {
    const g = makeGraph();
    const boot = g.bootstrap(new ActorId("system"));
    g.start();

    g.record(
      new EventType("test.event"), new ActorId("alice"), {},
      [boot.id], new ConversationId("conv_1"),
    );
    g.record(
      new EventType("test.event"), new ActorId("bob"), {},
      [boot.id], new ConversationId("conv_1"),
    );

    const results = g.query().bySource(new ActorId("alice"), 10);
    expect(results.length).toBe(1);
    expect(results[0].source.value).toBe("alice");
  });

  it("query.eventCount reflects store state", () => {
    const g = makeGraph();
    g.bootstrap(new ActorId("system"));
    g.start();
    expect(g.query().eventCount()).toBe(1);
  });

  it("query.actor retrieves registered actors", () => {
    const g = makeGraph();
    g.bootstrap(new ActorId("system"));
    g.start();

    const { publicKey } = makeActor("carol");
    const registered = g.actorStore.register(publicKey, "Carol", ActorType.Human);

    const found = g.query().actor(registered.id);
    expect(found.displayName).toBe("Carol");
  });

  it("close is idempotent", () => {
    const g = makeGraph();
    g.bootstrap(new ActorId("system"));
    g.start();
    g.close();
    g.close(); // second close should not throw
    expect(() => g.start()).toThrow(GraphStateError);
  });

  it("custom signer is used for bootstrap and record", () => {
    const calls: number[] = [];
    const spy: import("../src/event.js").Signer = {
      sign(data: Uint8Array) {
        calls.push(data.length);
        return new NoopSigner().sign(data);
      },
    };

    const g = makeGraph({ signer: spy });
    g.bootstrap(new ActorId("system"));
    expect(calls.length).toBe(1);

    g.start();
    const boot = g.store.head().unwrap();
    g.record(
      new EventType("test.event"), new ActorId("alice"), {},
      [boot.id], new ConversationId("conv_1"),
    );
    expect(calls.length).toBe(2);
  });

  it("record with per-call signer overrides default", () => {
    const defaultCalls: number[] = [];
    const overrideCalls: number[] = [];
    const defaultSigner: import("../src/event.js").Signer = {
      sign(data: Uint8Array) {
        defaultCalls.push(data.length);
        return new NoopSigner().sign(data);
      },
    };
    const overrideSigner: import("../src/event.js").Signer = {
      sign(data: Uint8Array) {
        overrideCalls.push(data.length);
        return new NoopSigner().sign(data);
      },
    };

    const g = makeGraph({ signer: defaultSigner });
    g.bootstrap(new ActorId("system"));
    g.start();

    const boot = g.store.head().unwrap();
    g.record(
      new EventType("test.event"), new ActorId("alice"), {},
      [boot.id], new ConversationId("conv_1"), overrideSigner,
    );

    // Default was used for bootstrap, override for record
    expect(defaultCalls.length).toBe(1);
    expect(overrideCalls.length).toBe(1);
  });

  it("query.trustScore and trustBetween delegate to trust model", () => {
    const g = makeGraph();
    g.bootstrap(new ActorId("system"));
    g.start();

    const { publicKey: pk1 } = makeActor("dave");
    const { publicKey: pk2 } = makeActor("eve\0");
    const a1 = g.actorStore.register(pk1, "Dave", ActorType.Human);
    const a2 = g.actorStore.register(pk2, "Eve", ActorType.AI);

    const score = g.query().trustScore(a1);
    expect(score.overall.value).toBeGreaterThanOrEqual(0);
    expect(score.overall.value).toBeLessThanOrEqual(1);

    const between = g.query().trustBetween(a1, a2);
    expect(between.overall.value).toBeGreaterThanOrEqual(0);
  });

  it("query.byConversation filters correctly", () => {
    const g = makeGraph();
    const boot = g.bootstrap(new ActorId("system"));
    g.start();

    g.record(
      new EventType("test.event"), new ActorId("alice"), {},
      [boot.id], new ConversationId("conv_a"),
    );
    g.record(
      new EventType("test.event"), new ActorId("alice"), {},
      [boot.id], new ConversationId("conv_b"),
    );

    const results = g.query().byConversation(new ConversationId("conv_a"), 10);
    expect(results.length).toBe(1);
    expect(results[0].conversationId.value).toBe("conv_a");
  });
});

File: rust/tests/graph_test.rs
use std::collections::BTreeMap;

use serde_json::Value;

use eventgraph::actor::{Actor, ActorStatus, ActorStore, ActorType, InMemoryActorStore};
use eventgraph::decision::AuthorityLevel;
use eventgraph::event::Signer;
use eventgraph::graph::{Graph, GraphConfig};
use eventgraph::store::{InMemoryStore, Store};
use eventgraph::types::{
    ActorId, ConversationId, EventType, PublicKey, Signature,
};

// ── Helpers ──────────────────────────────────────────────────────────

struct TestSigner;

impl Signer for TestSigner {
    fn sign(&self, data: &[u8]) -> Signature {
        let mut buf = vec![0u8; 64];
        let len = data.len().min(64);
        buf[..len].copy_from_slice(&data[..len]);
        Signature::new(buf).unwrap()
    }
}

fn test_actor(name: &str, key_byte: u8) -> Actor {
    let mut key = [0u8; 32];
    key[0] = key_byte;
    Actor::new(
        ActorId::new(name).unwrap(),
        PublicKey::new(key),
        name.to_string(),
        ActorType::Human,
        BTreeMap::new(),
        1_000_000_000,
        ActorStatus::Active,
    )
}

fn new_test_graph() -> (Graph, ActorId) {
    let store = InMemoryStore::new();
    let actor_store = InMemoryActorStore::new();
    let g = Graph::new(store, actor_store);
    g.start().expect("start should succeed");
    let actor_id = ActorId::new("actor_system0000000000000000000001").unwrap();
    (g, actor_id)
}

// ── 1. TestNewGraph ─────────────────────────────────────────────────

#[test]
fn test_new_graph() {
    let store = InMemoryStore::new();
    let actor_store = InMemoryActorStore::new();
    let g = Graph::new(store, actor_store);

    assert_eq!(g.store().count(), 0);
    assert_eq!(g.actor_store().actor_count(), 0);
}

// ── 2. TestBootstrap ────────────────────────────────────────────────

#[test]
fn test_bootstrap() {
    let (mut g, actor_id) = new_test_graph();

    let ev = g.bootstrap(actor_id.clone(), Some(&TestSigner)).unwrap();
    assert_eq!(ev.event_type.value(), "system.bootstrapped");
    assert_eq!(ev.source.value(), actor_id.value());
    assert_eq!(g.store().count(), 1);
}

// ── 3. TestBootstrapDoubleReject ────────────────────────────────────

#[test]
fn test_bootstrap_double_reject() {
    let (mut g, actor_id) = new_test_graph();

    g.bootstrap(actor_id.clone(), Some(&TestSigner)).unwrap();
    let result = g.bootstrap(actor_id, Some(&TestSigner));
    assert!(result.is_err(), "second bootstrap should fail");
}

// ── 4. TestRecord ───────────────────────────────────────────────────

#[test]
fn test_record() {
    let (mut g, actor_id) = new_test_graph();

    let bootstrap = g.bootstrap(actor_id.clone(), Some(&TestSigner)).unwrap();

    let mut content = BTreeMap::new();
    content.insert("key".to_string(), Value::String("value".to_string()));

    let ev = g
        .record(
            EventType::new("trust.updated").unwrap(),
            actor_id,
            content,
            vec![bootstrap.id.clone()],
            ConversationId::new("conv_test000000000000000000000001").unwrap(),
            Some(&TestSigner),
        )
        .unwrap();

    assert_eq!(ev.event_type.value(), "trust.updated");
    assert_eq!(g.store().count(), 2);
}

// ── 5. TestRecordAfterClose ─────────────────────────────────────────

#[test]
fn test_record_after_close() {
    let (mut g, actor_id) = new_test_graph();

    let bootstrap = g.bootstrap(actor_id.clone(), Some(&TestSigner)).unwrap();
    g.close();

    let result = g.record(
        EventType::new("trust.updated").unwrap(),
        actor_id,
        BTreeMap::new(),
        vec![bootstrap.id.clone()],
        ConversationId::new("conv_test000000000000000000000001").unwrap(),
        Some(&TestSigner),
    );
    assert!(result.is_err(), "record after close should fail");
}

// ── 6. TestEvaluate ─────────────────────────────────────────────────

#[test]
fn test_evaluate() {
    let (g, _actor_id) = new_test_graph();

    let alice = test_actor("alice", 1);
    let result = g.evaluate(&alice, "test.action").unwrap();

    // Default authority chain returns Notification for unknown actions
    assert_eq!(result.level, AuthorityLevel::Notification);
}

// ── 7. TestEvaluateAfterClose ───────────────────────────────────────

#[test]
fn test_evaluate_after_close() {
    let (mut g, _actor_id) = new_test_graph();
    g.close();

    let alice = test_actor("alice", 1);
    let result = g.evaluate(&alice, "test.action");
    assert!(result.is_err(), "evaluate after close should fail");
}

// ── 8. TestQuery ────────────────────────────────────────────────────

#[test]
fn test_query_event_count_and_recent() {
    let (mut g, actor_id) = new_test_graph();

    g.bootstrap(actor_id, Some(&TestSigner)).unwrap();

    let q = g.query().unwrap();
    assert_eq!(q.event_count(), 1);

    let recent = q.recent(10);
    assert_eq!(recent.len(), 1);
    assert_eq!(recent[0].event_type.value(), "system.bootstrapped");
}

// ── 9. TestQueryByType ──────────────────────────────────────────────

#[test]
fn test_query_by_type() {
    let (mut g, actor_id) = new_test_graph();
    g.bootstrap(actor_id, Some(&TestSigner)).unwrap();

    let q = g.query().unwrap();
    let et = EventType::new("system.bootstrapped").unwrap();
    let events = q.by_type(&et, 10);
    assert_eq!(events.len(), 1);
}

// ── 10. TestQueryBySource ───────────────────────────────────────────

#[test]
fn test_query_by_source() {
    let (mut g, actor_id) = new_test_graph();
    g.bootstrap(actor_id.clone(), Some(&TestSigner)).unwrap();

    let q = g.query().unwrap();
    let events = q.by_source(&actor_id, 10);
    assert_eq!(events.len(), 1);
}

// ── 11. TestStartIdempotent ─────────────────────────────────────────

#[test]
fn test_start_idempotent() {
    let store = InMemoryStore::new();
    let actor_store = InMemoryActorStore::new();
    let g = Graph::new(store, actor_store);

    g.start().unwrap();
    g.start().unwrap(); // second start should be a no-op
}

// ── 12. TestCloseIdempotent ─────────────────────────────────────────

#[test]
fn test_close_idempotent() {
    let store = InMemoryStore::new();
    let actor_store = InMemoryActorStore::new();
    let mut g = Graph::new(store, actor_store);

    g.close();
    g.close(); // second close should be a no-op
}

// ── 13. TestQueryAncestorsDescendants ───────────────────────────────

#[test]
fn test_query_ancestors_and_descendants() {
    let (mut g, actor_id) = new_test_graph();

    let bootstrap = g.bootstrap(actor_id.clone(), Some(&TestSigner)).unwrap();

    let mut content = BTreeMap::new();
    content.insert("key".to_string(), Value::String("val".to_string()));

    let child = g
        .record(
            EventType::new("trust.updated").unwrap(),
            actor_id,
            content,
            vec![bootstrap.id.clone()],
            ConversationId::new("conv_test000000000000000000000001").unwrap(),
            Some(&TestSigner),
        )
        .unwrap();

    let q = g.query().unwrap();

    // Ancestors of child should include bootstrap
    let ancestors = q.ancestors(&child.id, 10).unwrap();
    assert!(
        ancestors.iter().any(|e| e.id == bootstrap.id),
        "expected bootstrap in ancestors"
    );

    // Descendants of bootstrap should include child
    let descendants = q.descendants(&bootstrap.id, 10).unwrap();
    assert!(
        descendants.iter().any(|e| e.id == child.id),
        "expected child in descendants"
    );
}

// ── 14. TestQueryTrustScore ─────────────────────────────────────────

#[test]
fn test_query_trust_score() {
    let (g, _) = new_test_graph();

    let alice = test_actor("alice", 1);

    let q = g.query().unwrap();
    let metrics = q.trust_score(&alice).unwrap();
    assert_eq!(metrics.overall.value(), 0.0, "initial trust should be 0.0");
}

// ── 15. TestQueryTrustBetween ───────────────────────────────────────

#[test]
fn test_query_trust_between() {
    let (g, _) = new_test_graph();

    let alice = test_actor("alice", 1);
    let bob = test_actor("bob", 2);

    let q = g.query().unwrap();
    let metrics = q.trust_between(&alice, &bob).unwrap();
    assert_eq!(
        metrics.overall.value(),
        0.0,
        "initial trust between should be 0.0"
    );
}

// ── 16. TestQueryActor ──────────────────────────────────────────────

#[test]
fn test_query_actor() {
    let store = InMemoryStore::new();
    let mut actor_store = InMemoryActorStore::new();
    let pk = PublicKey::new([3u8; 32]);
    let registered = actor_store
        .register(pk, "Charlie", ActorType::Human)
        .unwrap();

    let g = Graph::new(store, actor_store);
    g.start().unwrap();

    let q = g.query().unwrap();
    let found = q.actor(registered.id()).unwrap();
    assert_eq!(found.id().value(), registered.id().value());
    assert_eq!(found.display_name(), "Charlie");
}

// ── 17. TestRecordBeforeStart ───────────────────────────────────────

#[test]
fn test_record_before_start() {
    let store = InMemoryStore::new();
    let actor_store = InMemoryActorStore::new();
    let mut g = Graph::new(store, actor_store);
    // Do NOT call start

    let actor_id = ActorId::new("actor_system0000000000000000000001").unwrap();
    let result = g.bootstrap(actor_id, Some(&TestSigner));
    assert!(result.is_err(), "bootstrap before start should fail");
}

// ── 18. TestWithCustomConfig ────────────────────────────────────────

#[test]
fn test_with_custom_config() {
    let store = InMemoryStore::new();
    let actor_store = InMemoryActorStore::new();
    let config = GraphConfig {
        subscriber_buffer_size: 512,
        fallback_to_mechanical: false,
    };
    let g = Graph::with_config(store, actor_store, config);
    assert_eq!(g.store().count(), 0);
}
